filesystem/mountpoint.go

```diff
@@ -123,20 +128,85 @@ func parseMountInfoLine(line string) *Mount {
 	if err != nil {
 		return nil
 	}
+	mnt.Subtree = unescapeString(fields[3])
 	mnt.Path = unescapeString(fields[4])
+	opts := strings.Split(fields[5], ",")
+	for _, opt := range opts {
+		if opt == "ro" {
+			mnt.ReadOnly = true
+		}
+	}
 	mnt.FilesystemType = unescapeString(fields[n+1])
 	mnt.Device = getDeviceName(mnt.DeviceNumber)
 	return mnt
 }
 
+func findMainMount(deviceMounts []*Mount) *Mount {
+	// Build a map mnt.Subtree => mnt for this filesystem's mounts.  If the
+	// same subtree is mounted multiple times, keep only one of its mounts,
+	// preferably a read-write one.
+	thisDevicesMountsBySubtree := make(map[string]*Mount)
+	for _, mnt := range deviceMounts {
+		otherMountOfSameSubtree, ok := thisDevicesMountsBySubtree[mnt.Subtree]
+		if !ok || (otherMountOfSameSubtree.ReadOnly && !mnt.ReadOnly) {
+			thisDevicesMountsBySubtree[mnt.Subtree] = mnt
+		}
+	}
+	// Find a mounted subtree which contains every other mounted subtree.
+	// This is usually the "/" subtree.  We don't support multiple
+	// uncontained subtrees, e.g. "/a" and "/b" mounted without "/" also
+	// mounted.  But if there is e.g. only "/a" we'll use that.
+	var mainMount *Mount
+	for subtree, mnt := range thisDevicesMountsBySubtree {
+		var containingMnt *Mount
+		for subtree != "/" && containingMnt == nil {
+			subtree = filepath.Dir(subtree)
+			containingMnt = thisDevicesMountsBySubtree[subtree]
+		}
+		if containingMnt != nil {
+			continue
+		}
+		if mainMount == nil {
+			mainMount = mnt
+			continue
+		}
+		log.Printf("Unsupported case: %q (%v) has multiple non-overlapping mounts. This filesystem will be ignored!",
+			mainMount.Device, mainMount.DeviceNumber)
+		return nil
+	}
+	return mainMount
+}
+
+// For each filesystem (i.e. for each unique device number), find the "main"
+// Mount for that filesystem, and keep only that one.  The "main" Mount is where
+// the fscrypt metadata is stored, so it's the only one we care about.
+func determineMainMounts() {
+	allMountsByDevice := make(map[DeviceNumber][]*Mount)
+	for _, mnt := range mountsByPath {
+		allMountsByDevice[mnt.DeviceNumber] =
+			append(allMountsByDevice[mnt.DeviceNumber], mnt)
+	}
+	for _, deviceMounts := range allMountsByDevice {
+		mainMount := findMainMount(deviceMounts)
+		for _, mnt := range deviceMounts {
+			if mnt != mainMount {
+				delete(mountsByPath, mnt.Path)
+			}
+		}
+		if mainMount != nil {
+			mountsByDevice[mainMount.DeviceNumber] = mainMount
+		}
+	}
+}
```

Comment:
Looking at all this, I'm not sure it's the right idea to try to support bind mounts without a root mount. The reason is that, for a given mountpoint, where we put the fscrypt metadata changes based on an unrelated mount point.

For example, if we have a bind mount with a `/a` subtree, then we will put fscrypt metadata at `/a`. Then if we do a normal mount of this filesystem somewhere, fscrypt metadata will now be put at `/`. I think the best policy should just be to always place fscrypt metadata at `/` and if you have a bind-mount without a corresponding normal mount, we just don't support the use-case.

This would also allow for code simplification, we would just want to be sure we gave proper errors. One way to do this might be to have `mountsByDevice` have `nil` values if the device is valid, but has no normal mountpoints. The code for `determineMainMounts` would then look like:

```go
for _, mnt := range mountsByPath {
	existingMnt, ok := mountsByDevice[mnt.DeviceNumber]
	if ok && existingMnt != nil {
		if existingMnt.ReadOnly && !mnt.BindMnt {
			mountsByDevice[mnt.DeviceNumber] = mnt
		}
	} else {
		if mnt.BindMnt {
			mountsByDevice[mnt.DeviceNumber] = nil
		} else {
			mountsByDevice[mnt.DeviceNumber] = mnt
		}
	}
}
```

What are your thoughts on this? I'm worried about people ending up with multiple copies of `.fscrypt` on their filesystem, leading to data-loss.

Reply:
I suppose that for simplicity we should do what you're suggesting for now.  However, if someone ever wants to use fscrypt in a container, then non-"/" subtree support will need to be added.

Reply:
Done.